### PR TITLE
Strings: noodlification

### PR DESCRIPTION
**Pull request by vhavlena**
Extension of the noodlification according to both sides of an equation. Still space for optimizations (e.g., simulation reduction of segments).

### CODE REVIEW

**Comment by martinhruska** [comment]
I will check it today or tommorow.

File: include/mata/nfa.hh [hunk 1]
```diff
@@ -958,6 +958,19 @@ Nfa uni(const Nfa &lhs, const Nfa &rhs);
...
+ * @brief Compute intersection of two NFAs with a possibility of using multiple epsilons.
+ *
+ * @param[in] lhs First NFA to compute intersection for.
+ * @param[in] rhs Second NFA to compute intersection for.
+ * @param[in] preserve_epsilon Whether to compute intersection preserving epsilon transitions.
+ * @param[in] epsilons Set of symbols to be considered as epsilons
+ * @param[out] prod_map Mapping of pairs of the original states (lhs_state, rhs_state) to new product states.
+ * @return NFA as a product of NFAs @p lhs and @p rhs with ε-transitions preserved.
+ */
+Nfa intersection_eps(const Nfa& lhs, const Nfa& rhs, bool preserve_epsilon, const std::set<Symbol>& epsilons,
```
Comment (martinhruska) [inline include/mata/nfa.hh]: I am not sure whether we want this here or we should move it to `strings` or `algorithms`

File: include/mata/nfa.hh [hunk 2]
```diff
@@ -973,6 +986,21 @@ Nfa intersection(const Nfa& lhs, const Nfa& rhs,
...
+ * Supports epsilon symbols when @p use_epsilon is set to true.
+ * @param[in] lhs First automaton to concatenate.
+ * @param[in] rhs Second automaton to concatenate.
+ * @param[in] epsilon Epsilon to be used co concatenation (provided @p use_epsilon is true)
+ * @param[in] use_epsilon Whether to concatenate over epsilon symbol.
+ * @param[out] lhs_result_states_map Map mapping lhs states to result states.
+ * @param[out] rhs_result_states_map Map mapping rhs states to result states.
+ * @return Concatenated automaton.
+ */
+Nfa concatenate_eps(const Nfa& lhs, const Nfa& rhs, const Symbol& epsilon, bool use_epsilon = false,
```
Comment (martinhruska) [inline include/mata/nfa.hh]: the same as above

**Review by martinhruska — APPROVED** [review]
I cannot evaluate correctnes of noodlification algorithms. Code quality looks good to me as well as code culture. My only note is that we may move the epsilon versions of algorithms from interface file.

File: include/mata/nfa.hh (hunk 1 above)
Comment (vhavlena) [reply include/mata/nfa.hh]: I don't have a strong opinion to that. If the others are not strictly against, I switch it to`algorithms` header file.

**Comment by vhavlena** [comment]
Thanks for the review. The correctness can be proven by modus exampli (it works on two examples implies it is indeed correct).

**Comment by martinhruska** [comment]
> Thanks for the review. The correctness can be proven by modus exampli (it works on two examples implies it is indeed correct).

Well, then I consider it verified.

**Comment by martinhruska** [comment]
Unfortunately, there are some conflicts now, probably they raise after merge of another PR. Could you resolve it please?

**Comment by vhavlena** [comment]
I resolved conflicts

File: include/mata/nfa.hh (hunk 1 above)
Comment (vhavlena) [reply include/mata/nfa.hh]: I moved the functions to `Algorithm` namespace

**Comment by jurajsic** [comment]
It looks fine to me, but I would probably change the type of arguments of `noodlify_for_equation` from `const AutRefSequence&` and `const AutPtrSequence&` to `const std::vector<std::shared_ptr<Nfa>>&` because returned noodles contain shared pointers and we also use only shared pointers in our solver.

**Comment by vhavlena** [comment]
Sure, updated with shared pointers now.

**Comment by martinhruska** [comment]
Well, it looks that everyone is satisfied with the changes. If noone objects, I am gonna merge it later today.

**Pull request merged**